crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -639,3 +655,301 @@ fn normalize_string(
+/// ```python
+/// def sparkle_sky():
+///     """Make a pretty sparkly sky.
+///     *       * ✨        *.    .
+///        *       *      ✨      .
+///        .  *      . ✨    * .  .
+///     """
+/// ```
+/// We can't compute the full indentation here since we don't know what the block indent of
+/// the doc comment will be yet and which we can only have added by formatting each line
+/// separately with a hard line break. This means we need to strip shared indentation from
+/// docstring while preserving the in-docstring bigger-than-suite-statement indentation. Example:
+/// ```python
+/// def f():
+///  """first line
+///  line a
+///     line b
+///  """
+/// ```
+/// The docstring indentation is 2, the block indents will change this to 4 (but we can't
+/// determine this at this point). The indentation of line a is 2, so we trim `  line a`
+/// to `line a`. For line b it's 5, so we trim it to `line b` and pad with 5-2=3 spaces to
+/// `   line b`. The closing quotes, being on their own line, are stripped get only the
+/// default indentation. Fully formatted:
+/// ```python
+/// def f():
+///    """first line
+///    line a
+///       line b
+///    """
+/// ```
+///
+/// Tabs are counted by padding them to the next multiple of 8 according to `str.expandtabs`. When
+/// we see indentation that contains a tab or any other none ascii-space whitespace we rewrite the
+/// string.
+///
+/// Additionally, if any line in the docstring has less indentation than the docstring
+/// (effectively a negative indentation wrt. to the current level), we pad all lines to the
+/// level of the docstring with spaces.
+/// ```python
+/// def f():
+///    """first line
+/// line a
+///    line b
+///      line c
+///    """
+/// ```
+/// Here line a is 3 columns negatively indented, so we pad all lines by an extra 3 spaces:
+/// ```python
+/// def f():
+///    """first line
+///    line a
+///       line b
+///         line c
+///    """
+/// ```
+fn format_docstring(
+    // The start of the string prefix to after the closing quotes
+    outer_range: TextRange,
+    f: &mut PyFormatter,
+) -> FormatResult<()> {
+    let locator = f.context().locator();
+    let string_part = FormatStringPart::new(
+        outer_range,
+        // It's not an f-string
+        Quoting::CanChange,
+        &locator,
+        f.options().quote_style(),
+    );
+
+    // Black doesn't change the indentation of docstrings that contain an escaped newline
+    if locator.slice(outer_range).contains("\\\n") {
+        return string_part.fmt(f);
+    }
+
+    let (normalized, _) = normalize_string(
+        locator.slice(string_part.range),
+        string_part.preferred_quotes,
+        string_part.is_raw_string,
+    );
+    // is_borrowed is unstable :/
+    let already_normalized = matches!(normalized, Cow::Borrowed(_));
+
+    let mut lines = normalized.lines().peekable();
+
+    // Start the string
+    write!(f, [string_part.prefix, string_part.preferred_quotes])?;
```

Comment:
Can we emit the source position before `string_part.prefix` and the end position at the very end. This will help getting correct source maps.